### PR TITLE
replace custom showdown fork with markdown-it

### LINKED ISSUE #1501: Hyperlinks containing underscores are processed as italics
Hyperlinks that contain matched pairs of underscores are converted to italics, which breaks the hyperlink. This affects both the Live Preview editor and published content.

Ghost Version: 0.3.3
### Example

Enter the following URL in the Ghost editor:

```
http://en.wikipedia.org/wiki/Tourism_in_Germany
```
#### Expected HTML output

```
<a href="http://en.wikipedia.org/wiki/Tourism_in_Germany">http://en.wikipedia.org/wiki/Tourism_in_Germany</a>
```
#### Actual HTML output

```
<a href="http://en.wikipedia.org/wiki/Tourism">http://en.wikipedia.org/wiki/Tourism</a><em>in</em>Germany
```

2013-11-17_12h36_19

### PR DESCRIPTION
**Pull request by kevinansfield**
refs https://github.com/TryGhost/Ghost-Admin/pull/690
closes #1501, #2093, #4592, #4627,  #4659, #5039, #5237, #5587, #5625, #5632, #5822, #5939, #6840, #7183, #7536

- swaps showdown for markdown-it when rendering markdown

This has a number of breaking changes, so far the following have been identified:
- ~~headings must have a space after the `#`, e.g. `#test` will output as plain text~~~
- ~~inline \`code\` is wrapped in `<code>` when inside of a `<pre>` tag~~ (I think this is actually desired behaviour)
- ~~doesn't add ID's to headings~~
- uses `<s>` for strikethrough instead of `<del>`
- `'Foo'` is picked up as an image rather than a footnote
- `'====test==test==test====test'` results in nested `<mark>` tags

TODO:
- [x] see if any of the above are easily fixable via config
- [x] test against open showdown issues
- [x] determine which breaking changes are acceptable
- [x] ~~update failing tests to match new markup specs~~ removed specs so we will now rely on markdown-it's own test suite
- [x] remove duplication of markdown parser setup (move it into a util?)

<details>
<summary>Markdown used for testing open showdown issues:</summary>

    1501 underscores in links converted to italics http://en.wikipedia.org/wiki/Tourism_in_Germany

    2093 linebreak not working correctly
    1
    2
    *
    4
    #
    5
    :
    6

    4592 multiple underscores in a code block
    ```
    ____
    &#95;&#95;&#95;&#95;
    ```

    4627 double equals in URL converted to \<mark\>
    

    4659 whitespace at the end lines when copying code blocks
    ```
    sudo apt-get install \
    python2.7 \
    ```

    5039 md inside code fences fails
    ```
    ![m'lady](http://)
    ~~strike~~
    test
    ```

    5237 inline code in footnote link title
    Here is a footnote reference,[^1] and another.[^longnote]
    [^1]: [`Something`](http://example.com)

    5587 inline script has trouble with ==
    <script>
    var x = 0;
    if ( x === 0 ) {
        alert("yay");
    }
    </script>

    5625 image url with double slash doesn't render preview
    

    5632 fenced code block in list
    - a list item with a fenced code block
      ```javascript
      this 
      ```

    5822 editor crash
     [test](http.cat/404)

    5939 pre and code in fenced code block
    ```
    <pre>test</pre>
    <code>test</code>
    ```

    6840 code in footnotes
    This is normal text.[^1]

    `this` is a single line of code [^2]

    [^1]: this is still ok but `this` causes an error
    [^2]: <code>this</code> will also go crazy

    7183 images in numbered lists
    1. Sample content
    
    2. Second sample content
    
    3. Third sample content
    

    7536 backticks break code block rendering
    `` code+` ``

    ```second code block```

    `third code block`

</details>

### CODE REVIEW

**Comment by kirrg001** [comment]
I know this is WIP, but just leaving one thought here: If we replace showdown by markdown-it, should we ensure that all open showdown bugs are not present?

**Comment by kevinansfield** [comment]
Yep, my plan is to review those quickly tomorrow. I think the remaining breaking changes that the tests highlight are pretty edge case so probably a good point to check we're actually moving in the right direction 😅

**Comment by kevinansfield** [comment]
Tested against all open [issues labeled with `showdown`](https://github.com/tryghost/ghost/issues?utf8=✓&q=is%3Aissue%20label%3Ashowdown%20) and verified that the renderer switch fixes all of them 💃

**Comment by kevinansfield** [comment]
@kirrg001 would you be happy deleting the `core/test/unit/markdown_client_integrated_spec.js` file? It was testing our old custom Showdown extensions, it seems now that we're using `markdown-it` directly then we should be relying on their tests and to open issues on that project if there are rendering bugs.

**Comment by kirrg001** [comment]
> Tested against all open issues labeled with showdown and verified that the renderer switch fixes all of them 💃

wow fantastic 😎

**Comment by kirrg001** [comment]
> closes #1501, #2093, #4592, #4627, #4659, #5039, #5237, #5587, #5625, #5632, #5822, #5939, #6840, #7183, #7536

Is the plan to use markdown-it in LTS? Are we sure we want to close these bugs?

**Comment by kirrg001** [comment]
> @kirrg001 would you be happy deleting the core/test/unit/markdown_client_integrated_spec.js file? It was testing our old custom Showdown extensions, it seems now that we're using markdown-it directly then we should be relying on their tests and to open issues on that project if there are rendering bugs.

I had a quick look [at their test coverage](https://github.com/markdown-it/markdown-it/tree/master/test), doesn't look that impressive (in [comparison to showdown](https://github.com/showdownjs/showdown/tree/master/test)). I think if we can ensure that our needs are covered, i'm absolutely okay with removing this test file.

**Comment by kevinansfield** [comment]
> Is the plan to use markdown-it in LTS? Are we sure we want to close these bugs?

Not to my knowledge, the idea behind switching now is because there will be minor breaking changes in the rendering which is fine for 1.0. I'm not aware there are any plans to update showdown or resolve markdown rendering issues in LTS?

> I had a quick look at their test coverage, doesn't look that impressive (in comparison to showdown). I think if we can ensure that our needs are covered, i'm absolutely okay with removing this test file.

The majority of the coverage comes under the [CommonMark specs](https://github.com/markdown-it/markdown-it/blob/master/test/fixtures/commonmark/good.txt) which seem pretty comprehensive to me? They have a test harness that loads the spec files and runs the examples/compares output for every example in there, there are then additional tests for options which deviate from CommonMark.

**Comment by kirrg001** [comment]
> The majority of the coverage comes under the CommonMark specs which seem pretty comprehensive to me? They have a test harness that loads the spec files and runs the examples/compares output for every example in there, there are then additional tests for options which deviate from CommonMark.

👍

File: core/server/utils/index.js [hunk 1]
```diff
@@ -112,7 +112,8 @@ utils = {
     tokens: require('./tokens'),
     sequence: require('./sequence'),
     ghostVersion: require('./ghost-version'),
-    mobiledocConverter: require('./mobiledoc-converter')
+    mobiledocConverter: require('./mobiledoc-converter'),
+    markdownConverter: require('./markdown-converter')
```
Comment (kevinansfield) [inline core/server/utils/index.js]: I wasn't sure if this is needed as it's not used via `utils.markdownConverter` anywhere but I followed the existing `mobiledocConverter` require (even though that's not used via `utils.mobiledocConverter`)

File: core/server/models/post.js [hunk 2]
```diff
@@ -237,7 +237,7 @@ Post = ghostBookshelf.Model.extend({
             this.set('html', legacyConverter.render(_.toString(this.get('markdown'))));
         }
 
-        if (this.hasChanged('html')) {
+        if (this.hasChanged('html') || !this.get('plaintext')) {
```
Comment (kevinansfield) [inline core/server/models/post.js]: I'm not sure why the [test for this](https://github.com/TryGhost/Ghost/blob/master/core/test/integration/model/model_posts_spec.js#L433-L448) wasn't failing before or why it started to with this PR as the test doesn't change the HTML

**Comment by kirrg001** [comment]
There are 18 open showdown bugs (see https://github.com/TryGhost/Ghost/issues?q=is%3Aopen+is%3Aissue+label%3Ashowdown). You have marked 15 for closing. Is there any reason? If there is a reason, maybe we can leave a note in the other 3 issues.

**Comment by kevinansfield** [comment]
One of the 18 wasn't showdown related so I've removed that label, the other 2 are issues that can be indirectly closed by the renderer switch rather than bugs that are directly fixed by this PR.

https://github.com/TryGhost/Ghost/issues/5028
https://github.com/TryGhost/Ghost/issues/5598

File: core/server/models/post.js [hunk 3]
```diff
@@ -234,11 +233,11 @@ Post = ghostBookshelf.Model.extend({
         if (mobiledoc) {
             this.set('html', utils.mobiledocConverter.render(JSON.parse(mobiledoc)));
         } else {
-            // legacy showdown mode
-            this.set('html', legacyConverter.makeHtml(_.toString(this.get('markdown'))));
+            // legacy markdown mode
+            this.set('html', legacyConverter.render(_.toString(this.get('markdown'))));
```
Comment (kirrg001) [inline core/server/models/post.js]: This [test](https://github.com/TryGhost/Ghost/blob/master/core/test/integration/model/model_posts_spec.js#L433) hasn't failed before, because the result of showdown and markdown-it is different.
Showdown wraps the whole markdown into an outer `<p>` tag and that's why `hasChanged('html')` was true for showdown, but false for markdown-it.

File: core/server/utils/index.js (hunk 1 above)
Comment (kirrg001) [inline core/server/utils/index.js]: that's fine, doesn't matter so much. If somebody wants to require `require('../utils')` he has access to `utils. markdownConverter`.

File: package.json [hunk 4]
```diff
@@ -62,6 +62,11 @@
     "knex": "0.13.0",
     "knex-migrator": "2.0.16",
     "lodash": "4.17.4",
+    "markdown-it": "8.3.1",
+    "markdown-it-footnote": "3.0.1",
+    "markdown-it-lazy-headers": "^0.1.3",
```
Comment (kirrg001) [inline package.json]: We usually pin versions strict on the server. I wonder...now that we have a yarn lock file and e.g. `markdown-it-lazy-headers 0.1.4` get's released, nobody would auto install this version until i update the dependencies in our yarn.lock. So maybe we can make version pinning a bit loose again (on `LTS` we hadn't used shrinkwrap and that's why it was imported to pin strict)

File: core/server/apps/default-cards/cards/markdown.js [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
     parser;
 
 module.exports = {
         name: 'card-markdown',
         type: 'dom',
         render(opts) {
             parser = new SimpleDom.HTMLParser(tokenizer, opts.env.dom, SimpleDom.voidMap);
-            return parser.parse('<div class="kg-card-markdown">' + converter.makeHtml(opts.payload.markdown || '') + '</div>');
+            return parser.parse(''
+                + '<div class="kg-card-markdown">'
```
Comment (kirrg001) [inline core/server/apps/default-cards/cards/markdown.js]: Do we want to keep these `kg` name pattern?

File: package.json (hunk 4 above)
Comment (kevinansfield) [inline package.json]: Oops, I must have missed the `--exact` flag when adding that one. Agree that looser version numbers make sense now that we have `yarn.lock` - `ember-cli` will use the locked versions and we still generate a `shrinkwrap.json` for the release too.

I'll pin for this PR and we can discuss a wider change elsewhere 😄

File: core/server/apps/default-cards/cards/markdown.js (hunk 5 above)
Comment (kevinansfield) [inline core/server/apps/default-cards/cards/markdown.js]: I think so, we're still using the mobiledoc format and hopefully our koenig editor will be coming back in some form later on. If we get rid of the card classes now then there will be another breaking change post 1.0 when we expand mobiledoc support again.

Comment (kirrg001) [inline core/server/apps/default-cards/cards/markdown.js]: absolutely fine with me 👍

**Review by kirrg001 — APPROVED** [review]
Looks good to me. I've tested a bit and looked for any potential problems. It does not require to reset your database. I would merge it and we'll see if we find some issues 👍

File: core/server/utils/markdown-converter.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    converter = new MarkdownIt({
+        html: true,
+        breaks: true,
+        linkify: true
+    })
+    .use(require('markdown-it-footnote'))
+    .use(require('markdown-it-lazy-headers'))
+    .use(require('markdown-it-mark'))
+    .use(require('markdown-it-named-headers'), {
+        // match legacy Showdown IDs otherwise default is github style dasherized
```
Comment (kevinansfield) [inline core/server/utils/markdown-converter.js]: `markdown-it-named-headers` adds IDs to header elements so `# my header` becomes `<h1 id="my-header">my header</h1>`

The dasherized `my-header` is what GitHub Flavoured Markdown (markdown-it, current showdown, etc) outputs by default but our old Showdown renderer didn't dasherize and would instead output `id="myheader"`.

Rather than add a breaking change that would affect links to migrated content I thought it would be better to replicate the old behaviour.
